--- a/test.c
+++ b/test.c
@@ -807,19 +739,6 @@ send_tlp(TLPQuadWord *tlp, int tlp_len)
 			}
 			DEBUG_PRINTF("\n");
 		}
-		
-		/*uint32_t expected_data = bswap32(expected_dword[3]);*/
-		/*uint32_t actual_data = bswap32(tlp_dword[3]);*/
-		uint32_t expected_data = expected_dword[3];
-		uint32_t actual_data = tlp_dword[3];
-		for (i = 0; i < 31; ++i) {
-			int expected_bit = (expected_data >> i) & 1;
-			int actual_bit = (actual_data >> i) & 1;
-			if (expected_bit != actual_bit) {
-				DEBUG_PRINTF("Data bit %03d: Exp - %d; Act - %d\n",
-					i, expected_bit, actual_bit);
-			}
-		}
 		print_last_recvd_packet_ids();
 		return -1;
 	}
