--- a/hw/net/e1000e_core.c
+++ b/hw/net/e1000e_core.c
@@ -3370,33 +3362,27 @@ static const uint32_t mac_reg_init[] = {
     [TDFTS]   = 0x600,
     [POEMB]   = 0x30D,
     [PBS]     = 0x028,
-	[MANC]    = E1000_MANC_RMCP_EN,
+    [MANC]    = E1000_MANC_DIS_IP_CHK_ARP,
     [FACTPS]  = E1000_FACTPS_LAN0_ON | 0x20000000,
-    [SWSM]    = 0,
+    [SWSM]    = 1,
     [RXCSUM]  = E1000_RXCSUM_IPOFLD | E1000_RXCSUM_TUOFLD,
     [ITR]     = _E1000E_MIN_XITR,
     [EITR...EITR + E1000E_MSIX_VEC_NUM - 1] = _E1000E_MIN_XITR,
 };
 
-static void
-_e1000e_core_initialize_regs(E1000ECore *core)
-{
-    memset(core->phy, 0, sizeof core->phy);
-    memmove(core->phy, phy_reg_init, sizeof phy_reg_init);
-	memset(core->mac, 0, sizeof core->mac);
-	memmove(core->mac, mac_reg_init, sizeof mac_reg_init);
-}
-
 void
 e1000e_core_reset(E1000ECore *core)
 {
     int i;
 
-	timer_del(core->autoneg_timer);
+    timer_del(core->autoneg_timer);
 
     _e1000e_intrmgr_reset(core);
 
-	_e1000e_core_initialize_regs(core);
+    memset(core->phy, 0, sizeof core->phy);
+    memmove(core->phy, phy_reg_init, sizeof phy_reg_init);
+    memset(core->mac, 0, sizeof core->mac);
+    memmove(core->mac, mac_reg_init, sizeof mac_reg_init);
 
     core->rxbuf_min_shift = 1 + E1000_RING_DESC_LEN_SHIFT;
 
diff --git a/hw/pci/pci.c b/hw/pci/pci.c
index c770814..e269975 100644
